kernel/comps/virtio/src/device/socket/device.rs

```diff
@@ -0,0 +1,347 @@
+            let token = recv_queue.add_buf(&[], &[rx_buffer.buf_mut()])?;
+            assert_eq!(i, token);
+            assert_eq!(rx_buffers.put(rx_buffer) as u16, i);
+        }
+
+        if recv_queue.should_notify() {
+            debug!("notify receive queue");
+            recv_queue.notify();
+        }
+
+        let mut device = Self {
+            config: virtio_vsock_config.read().unwrap(),
+            guest_cid,
+            send_queue,
+            recv_queue,
+            event_queue,
+            rx_buffers,
+            transport,
+            callbacks: Vec::new(),
+        };
+
+        // Interrupt handler if vsock device config space changes
+        fn config_space_change(_: &TrapFrame) {
+            debug!("vsock device config space change");
+        }
+
+        // Interrupt handler if vsock device receives some packet.
+        // TODO: This will be handled by vsock socket layer.
+        fn handle_vsock_event(_: &TrapFrame) {
+            debug!("Packet received. This will be solved by socket layer");
+        }
+
+        device
+            .transport
+            .register_cfg_callback(Box::new(config_space_change))
+            .unwrap();
+        device
+            .transport
+            .register_queue_callback(QUEUE_RECV, Box::new(handle_vsock_event), false)
+            .unwrap();
+
+        device.transport.finish_init();
+
+        register_device(
+            super::DEVICE_NAME.to_string(),
+            Arc::new(SpinLock::new(device)),
+        );
+
+        Ok(())
+    }
+
+    /// Returns the CID which has been assigned to this guest.
+    pub fn guest_cid(&self) -> u64 {
+        self.guest_cid
+    }
+
+    /// Sends a request to connect to the given destination.
+    ///
+    /// This returns as soon as the request is sent; you should wait until `poll` returns a
+    /// [`VsockEventType::Connected`] event indicating that the peer has accepted the connection
+    /// before sending data.
+    pub fn connect(&mut self, connection_info: &ConnectionInfo) -> Result<(), SocketError> {
+        let header = VirtioVsockHdr {
+            op: VirtioVsockOp::Request as u16,
+            ..connection_info.new_header(self.guest_cid)
+        };
+        // Sends a header only packet to the TX queue to connect the device to the listening socket
+        // at the given destination.
+        self.send_packet_to_tx_queue(&header, &[])
+    }
+
+    /// Accepts the given connection from a peer.
+    pub fn accept(&mut self, connection_info: &ConnectionInfo) -> Result<(), SocketError> {
+        let header = VirtioVsockHdr {
+            op: VirtioVsockOp::Response as u16,
+            ..connection_info.new_header(self.guest_cid)
+        };
+        self.send_packet_to_tx_queue(&header, &[])
+    }
+
+    /// Requests the peer to send us a credit update for the given connection.
+    fn request_credit(&mut self, connection_info: &ConnectionInfo) -> Result<(), SocketError> {
```

Comment:
Can `request_credit` be initiated by user? There seems to be no corresponding socket operations (bind, connect, listen, shutdown, etc)

Reply:
At the beginning of `send`, the socket will first check whether peer has sufficient buffer space via requesting the credit. 

> If there is insufficient buffer space, the sender waits until virtqueue buffers are returned and checks buf_alloc and fwd_cnt again.